### DISCUSSION TITLE
Next.js makes requests to /_next/*, what is that for?

### DISCUSSION THREAD

**Discussion by Rongeegee** [vercel/next.js#41047] [Help] [answered]
In my middleware.js of Next Auth, I have

```
export async function middleware(req) {
    const session = await getToken({ req, secret: process.env.NEXTAUTH_SECRET, })
    const in_authentication_page = req.url.includes("/auth/signin") || req.url.includes("/auth/register");
  
    if (session && in_authentication_page) {
      return NextResponse.redirect(process.env.NEXTAUTH_URL); 
    }
    if (!session && !in_authentication_page){
      return NextResponse.redirect(process.env.NEXTAUTH_URL +  '/auth/signin');
    } 
    return NextResponse.next();
}
```
The second if statement ` if (!session && !in_authentication_page)` when I am on login page, because Next.js makes requests to `/_next/*`, such as /_next/static/chunks/main.js?ts=1664508635053 and /_next/static/chunks/react-refresh.js?ts=1664508635053. I don't know what they are for. But this is clearly causing an issue on my signin page as I redirect those request to my signin page while they are meant to be finished and return a response.
I made a temporary fix by `if (!session && !in_authentication_page && req.url.indexOf("/_next") > -1)`. However, I don't think this should be a permanent solution, and I don't know what the permanent solution should be.

**Comment by icyJoseph** [comment, marked as answer]
These are the so called `_next` internals. You can read a little more about these here: https://nextjs.org/docs/messages/middleware-upgrade-guide#executing-middleware-on-internal-nextjs-requests

With the newest middleware, that introduced `_next` internals on middleware, you can also write a **_matcher_** so that the `middleware` only happens for specific paths, before **_matchers_**, the classic solution was similar to what you have:

```js
const PUBLIC_FILE = /\.(.*)$/;

// later inside the middleware
  const {pathname} = req.nextUrl;
  if (
            pathname.startsWith('/_next') || // exclude next.js internals
            pathname.startsWith('/api') || //  exclude all API routes
            pathname.startsWith('/static') || // exclude static files
            PUBLIC_FILE.test(pathname) // exclude all files in the public folder
        ) {
            return NextResponse.next();
        }
// otherwise do process the request
```

However, as said above, matchers are a possibility, which you can find documented here: https://nextjs.org/docs/advanced-features/middleware#matcher

```js
export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - static (static files)
     * - favicon.ico (favicon file)
     * - _next internal calls
     */
    '/((?!api|static|favicon.ico|_next).*)',
  ],
}
```

I'd suggest testing manually with various requests, like images, and find the behaviour that works best for you.
